File: lib/streak.ts
/**
 * Personal streak + hearts state machine (pure). This is the canonical spec
 * of the rules; the evaluate_user_streak RPC mirrors it day-for-day. See
 * FORMULAS.md §2. Keep the two in sync when either changes.
 */

export const MAX_HEARTS = 3;

export type StreakState = {
  /** Consecutive settled days. */
  streak: number;
  /** Best streak ever reached. */
  best: number;
  /** Remaining "streak freeze" hearts, 0–MAX_HEARTS. */
  hearts: number;
};

export type DayOutcome = {
  /** A completed workout was logged that day (routine or plan). */
  trained: boolean;
  /** That day had a scheduled session — a routine or a non-meal plan item. */
  requiredDay: boolean;
};

function clampHearts(hearts: number): number {
  if (!Number.isFinite(hearts)) return MAX_HEARTS;
  return Math.max(0, Math.min(MAX_HEARTS, Math.trunc(hearts)));
}

/**
 * Apply one settled day to the streak/hearts state.
 *
 * - Trained → streak +1, best updated, regain a heart (capped).
 * - Rest day (not required, not trained) → nothing changes; the streak is safe.
 * - Missed a required day with hearts left → spend one, streak frozen.
 * - Missed a required day at 0 hearts → streak resets to 0, hearts refill.
 */
export function nextStreakState(
  state: StreakState,
  day: DayOutcome,
): StreakState {
  const hearts = clampHearts(state.hearts);

  if (day.trained) {
    const streak = state.streak + 1;
    return {
      streak,
      best: Math.max(state.best, streak),
      hearts: Math.min(hearts + 1, MAX_HEARTS),
    };
  }

  if (!day.requiredDay) {
    return { ...state, hearts };
  }

  if (hearts > 0) {
    return { ...state, hearts: hearts - 1 };
  }

  return { streak: 0, best: state.best, hearts: MAX_HEARTS };
}

File: apps/api/internal/domain/streak/streak.go
// Package streak is the personal streak + hearts state machine
// (FORMULAS.md §2). A day is settled once, after it ends.
package streak

// MaxHearts is how many missed required days a streak survives.
const MaxHearts = 3

// State is what profiles stores: current_streak, best_streak, hearts.
type State struct {
	Streak int `json:"streak"`
	Best   int `json:"best"`
	Hearts int `json:"hearts"`
}

// Day is one settled day across every routine and active plan.
type Day struct {
	// Trained: any completed log that date, of any discipline.
	Trained bool `json:"trained"`
	// RequiredDay: a routine or any active plan scheduled a non-meal item.
	RequiredDay bool `json:"requiredDay"`
}

// Next applies one settled day:
//   - trained → streak+1, best updated, regain a heart (capped)
//   - rest day (not required, not trained) → unchanged
//   - missed a required day with hearts → spend one, streak frozen
//   - missed a required day at 0 hearts → streak 0, hearts refill
func Next(state State, day Day) State {
	hearts := min(max(state.Hearts, 0), MaxHearts)

	if day.Trained {
		streak := state.Streak + 1
		return State{Streak: streak, Best: max(state.Best, streak), Hearts: min(hearts+1, MaxHearts)}
	}
	if !day.RequiredDay {
		return State{Streak: state.Streak, Best: state.Best, Hearts: hearts}
	}
	if hearts > 0 {
		return State{Streak: state.Streak, Best: state.Best, Hearts: hearts - 1}
	}
	return State{Streak: 0, Best: state.Best, Hearts: MaxHearts}
}
